src/components/core/GeoChart.tsx

```diff
@@ -56,89 +77,219 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+        const defaultColorsNumber = DEFAULT_COLORS.length;
+        const segmentItemNames: IPushpinCategoryLegendItem[] = segmentItemsUniq.map(
+            (item: Execution.IResultHeaderItem, index: number): IPushpinCategoryLegendItem => {
+                const segmentItemName: string = getHeaderItemName(item) || EMPTY_SEGMENT_ITEM;
+                const color: string = DEFAULT_COLORS[index % defaultColorsNumber];
+                return {
+                    name: segmentItemName,
+                    legendIndex: index,
+                    color,
+                    isVisible: true,
+                };
+            },
+        );
+        return segmentItemNames;
+    }
+
+    private getGeoChartInnerOptions(geoData: IGeoData): IGeoChartInnerOptions {
+        const {
+            execution: { executionResult },
+        } = this.props;
+        const { segment } = geoData;
+        const attributeHeaderItems = getGeoAttributeHeaderItems(executionResult, geoData);
+        let segmentData: IPushpinCategoryLegendItem[] = [];
+        if (segment) {
+            const segmentAttributeHeaderItems = attributeHeaderItems[segment.index];
+            segmentData = this.getSegmentSeries(segmentAttributeHeaderItems);
+        }
+        return {
             geoData,
+            segmentData,
         };
-        return chartRenderer(chartProps);
-    };
+    }
 
-    private renderLegend = (geoData: IGeoData): React.ReactElement => {
+    private getLegendProps(): IGeoChartLegendRendererProps {
+        const { geoChartOptions } = this;
         const {
             config,
-            execution: { executionResult, executionResponse },
-            legendRenderer,
             legendPosition: position,
+            execution: { executionResult, executionResponse },
         } = this.props;
-        const { size, color, segment } = geoData;
-        const sizeFormat = size ? getFormatFromExecutionResponse(size.index, executionResponse) : undefined;
 
+        if (!geoChartOptions) {
+            return {
+                config,
+                position,
+                geoData: {},
+            };
+        }
+        const { geoData } = this.geoChartOptions;
+        const { size, color, segment } = geoData;
         const { colorData, sizeData } = calculateLegendData(executionResult, geoData);
-        const colorFormat = color
-            ? getFormatFromExecutionResponse(color.index, executionResponse)
-            : undefined;
-        const isColorLegendVisible = Boolean(colorData && !segment);
+        let legendProps: IGeoChartLegendRendererProps = { config, position, geoData };
 
-        const legendProps = {
-            config,
-            position,
-            sizeData,
-            sizeFormat,
-            colorData,
-            colorFormat,
-            isColorLegendVisible,
-        };
-        return legendRenderer(legendProps);
+        if (sizeData) {
+            const sizeLegendOptions = {
```

Comment:
You don't need to make this variable, it is used 1 time only.
Just return it directly
`return { ...legendProps, sizeData, sizeFormat };`

The same for `colorLegendOptions`, `segmentLegendOptions`

Reply:
The legendProps will append value that depends on condition, so cannot return it directly.
It should be: 
`if (sizeData) {
            const sizeFormat: string =  getFormatFromExecutionResponse(size.index, executionResponse);
            legendProps = { ...legendProps, sizeData, sizeFormat  };
        }

        if (colorData) {
            const colorFormat: string =  getFormatFromExecutionResponse(color.index, executionResponse);
            legendProps = { ...legendProps, colorData, colorFormat  };
        }

        if (segment && this.state.enabledLegendItems.length) {
            const { enabledLegendItems } = this.state;
            legendProps = { ...legendProps, segmentData: enabledLegendItems, onItemClick: this.onLegendItemClick  };
        }

        return legendProps;`

Reply:
my bad 
thank you :)